### PR TITLE
kubeadm: exclude control plane nodes from external LBs

### PR DESCRIPTION
**Pull request by neolit123**
**What this PR does / why we need it**:

Apply the label:
"node.kubernetes.io/exclude-from-external-load-balancers"

To control plane nodes to preserve backwards compatibility
with the legacy mode where "master" nodes were excluded from
LBs.

**Which issue(s) this PR fixes**:

xref https://github.com/kubernetes/kubeadm/issues/2375

**Special notes for your reviewer**:

the upgrade parts can be removed in 1.22.
the linked issue will track.

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: apply the "node.kubernetes.io/exclude-from-external-load-balancers" label on control plane nodes during "init", "join" and "upgrade" to preserve backwards compatibility with the lagacy LB mode where nodes labeled as "master" where excluded. To opt-out you can remove the label from a node. See #97543 and the linked KEP for more details.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/1143-node-role-labels
```

### CODE REVIEW

File: cmd/kubeadm/app/phases/markcontrolplane/markcontrolplane.go
Comment (pacoxu) [inline cmd/kubeadm/app/phases/markcontrolplane/markcontrolplane.go:29]: need we add `will be removed in 1.22` comments here?

Comment (neolit123) [reply cmd/kubeadm/app/phases/markcontrolplane/markcontrolplane.go:29]: it's actually for 1.24+.

i guess it's optional to remove this constant, but the TODO is linked bellow.

> https://github.com/kubernetes/kubeadm/issues/2200

Comment (neolit123) [reply cmd/kubeadm/app/phases/markcontrolplane/markcontrolplane.go:29]: the TODO about upgrades is the other linked issue.

File: cmd/kubeadm/app/constants/constants.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/constants/constants.go:235]: I will add a little bit more detail here, documenting that we are adding this label in order to ensure backward compatibility after a change landed somewhere else

File: cmd/kubeadm/app/phases/markcontrolplane/markcontrolplane.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/phases/markcontrolplane/markcontrolplane.go:33]: might be I'm missing something, but why are we adding a deprecated suffix for removing it down in `markControlPlaneNode`?

Comment (fabriziopandini) [reply cmd/kubeadm/app/phases/markcontrolplane/markcontrolplane.go:33]: If the goal is to make the message in MarkControlPlane to report (deprecated), I would prefer to have something, might be an `AddDeprecationNotice` func,   that manipulates `labelsToAdd` before calling fmt.Printf instead of having this UX concern spreading around the code base.

Comment (neolit123) [reply cmd/kubeadm/app/phases/markcontrolplane/markcontrolplane.go:33]: that's the other way of doing it.
will amend later today.

**Comment by neolit123** [comment]
@pacoxu @fabriziopandini updated.

**Review by pacoxu — APPROVED** [review]

**Comment by fabriziopandini** [comment]
/lgtm

@andrewsykim could you take a look at this?

**Comment by neolit123** [comment]
/hold cancel
the logic seems valid to me, we are doing this both on init and upgrade.
also the label matches the original PR https://github.com/kubernetes/kubernetes/pull/97543

**Pull request merged**
